contracts/signature/Eip712Signature.sol

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.8.0 <0.9.0;
+
+abstract contract EIP712Signature {
+    uint256 private nonce;
+
+    function eip712Nonce() public view returns (uint256 signed) {
+        return nonce;
+    }
+
+    function eip712BumpNonce() internal {
+        nonce = nonce + 1;
+    }
+
+    function eip712SignedBy() internal returns (address signer) {
+        if (msg.data.length >= 4 + 65) {
+            (
+                bytes calldata dataTrimmed,
+                uint8 v,
+                bytes32 r,
+                bytes32 s
+            ) = _splitSignature(msg.data);
+
+            bytes32 txHash = _transactionHash(msg.value, dataTrimmed, nonce);
+
+            signer = ecrecover(txHash, v, r, s);
+        }
+    }
+
+    function _splitSignature(
+        bytes calldata data
+    )
+        private
+        pure
+        returns (bytes calldata dataTrimmed, uint8 v, bytes32 r, bytes32 s)
+    {
+        uint256 length = data.length;
+        dataTrimmed = data[0:length - 65];
+        r = bytes32(data[length - 65:]);
+        s = bytes32(data[length - 33:]);
+        v = uint8(bytes1(data[length - 1:]));
+    }
+
+    function _transactionHash(
+        uint256 value,
+        bytes calldata data,
+        uint256 _nonce
+    ) private view returns (bytes32) {
+        uint256 chainId;
+        assembly {
```

Comment:
Why fallback to assembly here? You can use `block.chainid`.

Reply:
👌